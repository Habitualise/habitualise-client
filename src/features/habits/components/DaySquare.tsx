import {themeColors} from '@/theme';
import React from 'react';

import {View, StyleSheet} from 'react-native';
import {Text} from 'react-native-paper';
import {habitStatus} from '../types';

interface DaySquareProps {
  dayOfMonth: string;
  dayString: string;
  status: habitStatus;
}

export const DaySquare = ({dayOfMonth, status, dayString}: DaySquareProps) => {
  return (
    <View style={{flex: 1, alignItems: 'center'}}>
      <Text
        variant="labelSmall"
        style={[
          daySquareStyles.dayString,
          status === 'completed' && daySquareStyles.completedText,
          status === 'missed' && daySquareStyles.missedText,
        ]}>
        {dayString}
      </Text>
      <View
        style={[
          daySquareStyles.container,
          status === 'completed' && daySquareStyles.completedContainer,
          status === 'missed' && daySquareStyles.missedContainer,
        ]}>
        <Text
          style={[
            daySquareStyles.dayOfMonth,
            status === 'completed' && daySquareStyles.completedText,
            status === 'missed' && daySquareStyles.missedText,
          ]}>
          {dayOfMonth}
        </Text>
      </View>
    </View>
  );
};

export const daySquareStyles = StyleSheet.create({
  container: {
    width: 40,
    height: 40,
    borderRadius: 8,
    borderWidth: 1.5,
    justifyContent: 'center',
    alignItems: 'center',
    borderColor: themeColors.grey[500],
  },
  dayString: {
    marginBottom: 4,
    color: themeColors.grey[500],
    fontWeight: 'bold',
  },
  dayOfMonth: {
    color: themeColors.grey[500],
    fontWeight: 'bold',
  },
  completedContainer: {
    backgroundColor: themeColors.successLight,
    borderColor: themeColors.success,
  },
  completedText: {
    color: themeColors.success,
  },
  missedContainer: {
    backgroundColor: themeColors.errorContainerLight,
    borderColor: themeColors.errorLight,
  },
  missedText: {
    color: themeColors.errorLight,
  },
});
