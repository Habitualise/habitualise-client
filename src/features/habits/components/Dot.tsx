import React, {useMemo} from 'react';
import {ColorValue, StyleSheet, View} from 'react-native';

import {calculateDotColour} from '../utils/calculateDotColour';
import {themeColors} from '@app/theme';
import {HabitColor} from '@app/context/types';

interface DotProps {
  colorType: HabitColor;
  completed: boolean;
  consecutiveDaysCompleted: number;
}

export const Dot = (props: DotProps) => {
  const {colorType, completed, consecutiveDaysCompleted} = props;
  const dotColour = useMemo(
    () => calculateDotColour(colorType, consecutiveDaysCompleted),
    [colorType, consecutiveDaysCompleted],
  );

  return (
    <View
      style={[
        styles(dotColour).dot,
        // completed && styles(dotColour).skippedDot,
        !completed && styles(dotColour).missedDot,
      ]}
    />
  );
};

const styles = (dotColour: ColorValue) =>
  StyleSheet.create({
    dot: {
      width: 13,
      height: 13,
      borderRadius: 5,
      backgroundColor: dotColour,
    },
    // KEEP: in case we want to implement skipped dots later
    // skippedDot: {
    //   borderWidth: 3,
    //   borderColor: dotColour,
    //   backgroundColor: 'transparent',
    // },
    missedDot: {
      backgroundColor: themeColors.grey[300],
    },
  });
