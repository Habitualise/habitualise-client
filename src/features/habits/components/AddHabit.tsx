import {HabitIcon} from '@/components/HabitIcon';
import {ShadowModal} from '@/components/ShadowModal';
import {LABEL} from '@/language';
import {themeColors} from '@/theme';
import React, {useState} from 'react';
import {Pressable, StyleSheet, View} from 'react-native';

import {IconButton, Text, TextInput} from 'react-native-paper';
import {daySquareStyles} from './DaySquare';

interface AddHabitProps {
  setModalVisible: (value: boolean) => void;
}

const daysOfWeek = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];

export const AddHabit = ({setModalVisible}: AddHabitProps) => {
  const [chooseIconVisible, setChooseIconVisible] = useState(false);

  const [selectedDays, setSelectedDays] = useState({
    mon: true,
    tue: true,
    wed: true,
    thu: true,
    fri: true,
    sat: true,
    sun: true,
  });

  const toggleDay = (day: string) => {
    setSelectedDays(prevState => ({
      ...prevState,
      [day]: !prevState[day],
    }));
  };

  return (
    <>
      <ShadowModal
        modalVisible={chooseIconVisible}
        setModalVisible={setChooseIconVisible}
        greyOverlay={false}>
        <IconButton
          icon="close"
          onPress={() => setChooseIconVisible(false)}
          style={styles.closeModalButton}
        />
      </ShadowModal>
      <IconButton
        icon="close"
        onPress={() => setModalVisible(false)}
        style={styles.closeModalButton}
      />
      <Text variant="titleMedium" style={styles.header}>
        {LABEL.ADD_HABIT}
      </Text>
      <View style={{flexDirection: 'row', alignItems: 'center'}}>
        <Pressable onPress={() => setChooseIconVisible(true)}>
          <HabitIcon
            iconName="image-search"
            colour={themeColors.primary}
            style={styles.chooseIconButton}
          />
        </Pressable>
        <TextInput label={LABEL.HABIT_NAME} style={styles.habitNameInput} />
      </View>
      <View style={styles.repeatContainer}>
        <Text variant="bodySmall" style={{marginTop: 16}}>
          {LABEL.REPEAT}
        </Text>
        <View style={styles.dayButtonContainer}>
          {daysOfWeek.map(day => (
            <Pressable onPress={() => toggleDay(day.toLowerCase())} key={day}>
              <View
                style={[
                  daySquareStyles.container,
                  selectedDays[day.toLowerCase()] &&
                    daySquareStyles.completedContainer,
                ]}>
                <Text
                  variant="bodySmall"
                  style={
                    selectedDays[day.toLowerCase()]
                      ? daySquareStyles.completedText
                      : styles.greyDayText
                  }>
                  {day}
                </Text>
              </View>
            </Pressable>
          ))}
        </View>
      </View>
    </>
  );
};

const styles = StyleSheet.create({
  greyDayText: {
    color: themeColors.grey[600],
  },
  dayButtonContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: 8,
  },
  repeatContainer: {
    marginTop: 16,
    width: '100%',
  },
  header: {
    marginBottom: 26,
  },
  closeModalButton: {
    position: 'absolute',
    top: 0,
    right: 0,
  },
  habitNameInput: {
    flex: 1,
  },
  chooseIconButton: {
    aspectRatio: 1,
    width: 55,
  },
});
