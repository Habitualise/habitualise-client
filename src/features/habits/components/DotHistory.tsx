import React from 'react';
import {View, StyleSheet} from 'react-native';

import {Dot} from './Dot';

interface HistoryDotViewProps {
  completionHistory: string[];
  colour: string;
}

export const DotHistory = (props: HistoryDotViewProps) => {
  const {completionHistory, colour} = props;

  const renderDots = () => {
    let streak = 0;
    if (!completionHistory) {
      return [];
    }
    return completionHistory.map((state, index) => {
      if (state === 'DONE') {
        streak += 1;
      } else if (state === 'MISSED') {
        streak = 0;
      }
      return (
        <Dot
          key={index}
          baseColour={colour}
          state={state}
          consecutiveDaysCompleted={streak}
        />
      );
    });
  };

  return (
    <View style={styles.container}>
      <View style={styles.row}>{renderDots().slice(0, 14)}</View>
      <View style={styles.row}>{renderDots().slice(14)}</View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'column',
    justifyContent: 'space-between',
    width: '100%',
  },
  row: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
});
