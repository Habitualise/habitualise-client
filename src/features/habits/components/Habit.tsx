import React from 'react';
import {Divider, IconButton, Text} from 'react-native-paper';

import {HabitIcon} from '@/components/HabitIcon';
import {StyleSheet, View} from 'react-native';
import {themeColors} from '@/theme';
import {LABEL} from '@/language';
import {WeekView} from './WeekView';

interface HabitProps {
  id: string;
  name: string;
  iconName: string;
  colour: string;
  completionPercentage: number;
  status: 'active' | 'archived';
  lastSevenCompletions: string[];
  daysDue: number[];
}

export const Habit = (props: HabitProps) => {
  const {
    id,
    name,
    iconName,
    colour,
    completionPercentage,
    status,
    lastSevenCompletions,
    daysDue,
  } = props;

  return (
    <>
      <View style={styles.cardContainer}>
        <HabitIcon iconName={iconName} colour={colour} />
        <View style={{flex: 1}}>
          <Text style={styles.habitName}>{name}</Text>
          <Text style={styles.habitDaysDue} variant={'bodySmall'}>
            {LABEL.DAYS_COMPLETED(completionPercentage)}
          </Text>
        </View>
        <IconButton
          icon={'dots-vertical'}
          onPress={() => console.log('three dots pressed')}
        />
      </View>
      <WeekView lastSevenCompletions={lastSevenCompletions} daysDue={daysDue} />
      <Divider />
    </>
  );
};

const styles = StyleSheet.create({
  cardContainer: {
    padding: 16,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  habitName: {
    marginBottom: 4,
    fontWeight: 'bold',
  },
  habitDaysDue: {
    color: themeColors.grey[500],
  },
});
