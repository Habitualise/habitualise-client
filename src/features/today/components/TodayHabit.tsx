import React from 'react';
import {Divider, Text} from 'react-native-paper';
import {StyleSheet, View} from 'react-native';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';

import {themeColors} from '@app/theme';
import {HabitIcon} from '@app/components/HabitIcon';
import {ACTIONS, useStore} from '@app/context/StoreContext';
import {formatDaysDue} from '../utils/formatDaysDue';
import {PressableCard} from '@app/components/PressableCard';
import {HabitColor} from '@app/context/types';

interface TodayHabitProps {
  id: string;
  name: string;
  iconName: string;
  colour: HabitColor;
  isCompletedToday: boolean;
  daysDue: number[];
}

export const TodayHabit = (props: TodayHabitProps) => {
  const {id, name, iconName, colour, isCompletedToday, daysDue} = props;
  const {dispatch} = useStore();

  const toggleHabitCompleted = () => {
    dispatch({
      type: ACTIONS.TOGGLE_HABIT_IS_COMPLETED,
      payload: id,
    });
  };

  return (
    <>
      <PressableCard
        onPress={toggleHabitCompleted}
        isGreyedOut={isCompletedToday}>
        <HabitIcon iconName={iconName} colour={colour} />
        <View style={{flex: 1}}>
          <Text style={styles(isCompletedToday).habitName}>{name}</Text>
          <Text
            style={styles(isCompletedToday).habitDaysDue}
            variant={'bodySmall'}>
            {formatDaysDue(daysDue)}
          </Text>
        </View>
        <MaterialCommunityIcons
          name={isCompletedToday ? 'check-circle' : 'circle-outline'}
          color={
            isCompletedToday
              ? themeColors.successGreyedOut
              : themeColors.grey[400]
          }
          size={25}
        />
      </PressableCard>
      <Divider />
    </>
  );
};

const styles = (isCompletedToday = false) =>
  StyleSheet.create({
    habitName: {
      marginBottom: 4,
      fontWeight: 'bold',
      textDecorationLine: isCompletedToday ? 'line-through' : 'none',
      color: isCompletedToday ? themeColors.grey[500] : themeColors.grey[900],
    },
    habitDaysDue: {
      color: isCompletedToday ? themeColors.grey[400] : themeColors.grey[500],
    },
  });
