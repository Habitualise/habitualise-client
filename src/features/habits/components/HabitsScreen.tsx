import React, {useState} from 'react';
import {SafeAreaView} from 'react-native-safe-area-context';
import {ScrollView, StyleSheet, View} from 'react-native';
import {Button, Divider, IconButton, Text} from 'react-native-paper';

import PaperView from '@app/components/PaperView';
import {LABEL} from '@app/language';
import {useStore} from '@app/context/StoreContext';
import {Habit} from './Habit';
import {themeColors} from '@app/theme';
// import {AddHabit} from './AddHabit';
// import {ShadowModal} from '@app/components/ShadowModal';

export const HabitsScreen = () => {
  const [habitsStatus, setHabitsStatus] = useState<'active' | 'archived'>(
    'active',
  );
  const {state} = useStore();
  const {habits} = state;

  //   const [modalVisible, setModalVisible] = useState(false);

  const setActive = () => {
    setHabitsStatus('active');
  };

  const setArchived = () => {
    setHabitsStatus('archived');
  };

  return (
    <SafeAreaView
      style={{
        flex: 1,
      }}
      edges={['top', 'left', 'right']}>
      {/* <ShadowModal
        modalVisible={modalVisible}
        setModalVisible={setModalVisible}>
        <AddHabit setModalVisible={setModalVisible} />
      </ShadowModal> */}

      <PaperView style={{flex: 1}}>
        <View style={styles.headerContainer}>
          <Text variant="headlineLarge" style={styles.heading}>
            {LABEL.HABITS_HEADER}
          </Text>
          <View style={styles.buttonContainer}>
            <View style={styles.buttonGroup}>
              <Button
                mode={habitsStatus === 'active' ? 'contained-tonal' : 'text'}
                textColor={themeColors.primary}
                onPress={setActive}
                icon="check"
                style={{marginRight: 8}}>
                {LABEL.ACTIVE}
              </Button>
              <Button
                mode={habitsStatus === 'archived' ? 'contained-tonal' : 'text'}
                textColor={themeColors.primary}
                onPress={setArchived}
                icon="archive">
                {LABEL.ARCHIVED}
              </Button>
            </View>
            <IconButton
              icon="plus"
              iconColor={themeColors.primary}
              onPress={() => console.log('add habit pressed')}
            />
          </View>
        </View>
        <Divider />
        <ScrollView>
          {habits
            .filter(habit => habit.status === habitsStatus)
            .map(habit => (
              <Habit
                key={habit.id}
                id={habit.id}
                name={habit.name}
                iconName={habit.iconName}
                colour={habit.colour}
                completionPercentage={habit.completionPercentage}
                status={habit.status}
                completionHistory={habit.completionHistory}
                daysDue={habit.daysDue}
              />
            ))}
        </ScrollView>
      </PaperView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  headerContainer: {
    paddingHorizontal: 16,
    paddingVertical: 24,
  },
  heading: {
    marginBottom: 5,
    fontWeight: '600',
  },
  buttonContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  buttonGroup: {
    flexDirection: 'row',
  },
});
