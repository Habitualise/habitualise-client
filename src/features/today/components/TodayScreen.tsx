import React from 'react';
import {ScrollView, View} from 'react-native';
import {Divider, Text} from 'react-native-paper';
import {SafeAreaView} from 'react-native-safe-area-context';

import {LABEL} from '@app/language';
import getTodayDateString from '../utils/getTodayDateString';
import {TodayHabit} from './TodayHabit';
import {useStore} from '@app/context/StoreContext';
import PaperView from '@app/components/PaperView';
import {commonStyles} from '@app/components/styles';
import {getTodaysHabits} from '@app/features/today/utils/getTodaysHabits';
import Spinner from '@app/components/Spinner';

export const TodayScreen = () => {
  const {state} = useStore();
  const {loading} = state;
  const allHabits = state.habits;
  const todaysHabits = getTodaysHabits(allHabits);

  return (
    <SafeAreaView
      style={commonStyles.safeArea}
      edges={['top', 'left', 'right']}>
      <PaperView style={{flex: 1}}>
        <View style={commonStyles.headerContainer}>
          <Text variant="bodySmall">{getTodayDateString()}</Text>
          <Text variant="headlineLarge" style={commonStyles.heading}>
            {LABEL.TODAY_HEADER}
          </Text>
        </View>
        <Divider />
        <ScrollView>
          {loading && <Spinner />}
          {todaysHabits.map(habit => (
            <TodayHabit
              key={habit.id}
              id={habit.id}
              iconName={habit.iconName}
              name={habit.name}
              isCompletedToday={habit.isCompletedToday}
              daysDue={habit.daysDue}
              colour={habit.colour}
            />
          ))}
        </ScrollView>
      </PaperView>
    </SafeAreaView>
  );
};
