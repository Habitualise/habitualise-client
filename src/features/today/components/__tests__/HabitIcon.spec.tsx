import React from 'react';
import { render } from '@testing-library/react-native';

import { HabitIcon } from '@/components/HabitIcon';

describe('HabitIcon', () => {
  it('should match snapshot', () => {
    const { toJSON } = render(
      <HabitIcon colour="rgb(0, 0, 0" iconName="check" />,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
