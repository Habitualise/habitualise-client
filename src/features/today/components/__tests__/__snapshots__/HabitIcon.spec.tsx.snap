// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HabitIcon should match snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "rgba(0, 0, 0",
      "borderRadius": 50,
      "flexDirection": "row",
      "justifyContent": "center",
      "marginRight": 16,
      "padding": 5,
    }
  }
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      [
        {
          "color": "rgb(0, 0, 0",
          "fontSize": 32,
        },
        undefined,
        {
          "fontFamily": "Material Design Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        {},
      ]
    }
  >
    ó°„¬
  </Text>
</View>
`;
